### ISSUE TITLE
[Tracking] [External] Attendee Tracking project

### ISSUE THREAD

**Issue by Julesssss** [Expensify/App#44725]
# Important Links
- [Design Doc](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?pli=1#bookmark=id.xbwi13ou7vjv)
- [Pre-design](https://expensify.slack.com/archives/C05EQBQG6E5/p1699272711013409)
- [Internal tracking issue](https://github.com/Expensify/Expensify/issues/306605)

# Strategy
In order to jump start viral growth, we need to migrate all existing customers to NewDot as soon as possible, so they can reap the rewards of the new app we've been building for several years.

# Problem 
It's common that one employee incurs an expense that includes others (eg: one person buys lunch for their whole office). There's no way to clearly outline this in NewDot though, so it can easily look like they overspent. This can drive manual back and forth between employee and admin, and cause significant loss of efficiency when compounded across the organization. A less efficient experience adds friction to enticing members to make the switch to NewDot and will slow down reunification.

# Solution 
Add support for [attendee tracking](https://community.expensify.com/discussion/5570/deep-dive-what-is-attendee-tracking) in NewDot. This will make it easy for members to clearly indicate when spend is shared with others, and should reduce back and forth about larger than normal requests. The exact scope is TBD, but my thinking is we can do this by:

- Adding an Attendees row on requests in NewDot.
- Allowing members tap on the row, see a list of all eligible attendees, and select from the list.
- Surfacing selected attendees, and amount per person on request detailed.

![Screenshot 2024-07-18 at 14 49 04](https://github.com/user-attachments/assets/8115bd53-acb5-4f9f-9351-e141a50b53ab)

## Implementation Issues

**Frontend:**
- https://github.com/Expensify/App/issues/47171
- https://github.com/Expensify/App/issues/47172
- https://github.com/Expensify/App/issues/47173
- https://github.com/Expensify/App/issues/48695
- https://github.com/Expensify/App/issues/51335

**Backend**
- https://github.com/Expensify/Expensify/issues/425201
- https://github.com/Expensify/Expensify/issues/419205
- https://github.com/Expensify/Expensify/issues/419206
- https://github.com/Expensify/Expensify/issues/419218
- https://github.com/Expensify/App/issues/49326

#### TODO
**Finish the project**
- https://github.com/Expensify/Expensify/issues/429056
- https://github.com/Expensify/App/issues/48692
- Update UpdateMoneyRequestAttendees  command to return [missing](https://github.com/Expensify/Web-Expensify/actions/runs/11458759213/job/31881828321#step:16:140) onyx data: `_transactions`, `_report`, and `reportNameValuePairs`
- Within `createUpdateMoneyRequestAttendees`, update the users recent attendees and pass back to client
- App client should update the param used to pass attendees to backend
- App client should switch to the `transaction.attendees` onyx keys
- Unblock feature in client by reversing [this commit](https://github.com/Expensify/App/commit/e82ef4e9f7fced9dea35067c20c748d4db50852e)

**Post-launch polish:**
- Clear out uses of the now-unused Auth attendee param
- Prevent violation bug by returning OnyxData like done[ here](https://github.com/Expensify/Web-Expensify/pull/43937/files#diff-8367f109728fcc54db479ccc765a52d2115e52d35eba9c660e665be1a34411acL4166-R4171) for other updates
- Add more tests like testRequestMoneyPersistsAttendees for edit case
- https://github.com/Expensify/App/issues/50802
- Once fully migrated, introduce minimum attendee size of 1:
`Request::verifyAttributeSize(request, "attendees", 0, Request::MAX_SIZE_NONCOLUMN);`

## Tasks

- [ ] Post **Proposal (full Problem/Solution statement)** in `#expensify-open-source`
- [ ] Wait at least one full business day, and until the post has a majority (2/3) of positive reactions (👍)
- [ ] Paste Proposal in the space above with a link to the Slack thread
- [ ] Email `strategy@expensify.com` and paste in the Proposal
- [ ] Fill out the **High-level overview of the problem**, **Timeline**, and **Terminology** sections of the Design Doc
- [ ] Email `strategy@expensify.com` (continue the same email chain as before) with the link to your Design Doc
- [ ] Host a **pre-design** meeting ([example](https://expensify.slack.com/archives/C01GTK53T8Q/p1665794669539419)) in `#expensify-open-source` to discuss any necessary details in public before filling out the **High-level of proposed solution** section.
- [ ] Fill out the **High-level of proposed solution** section
- [ ] Email `stategy@expensify.com` again with links to the doc and pre-design conversation in Slack
- [ ] Add the `DesignDocReview` label to get the **High-level of proposed solution** section reviewed
- [ ] Respond to any questions or concerns and bring up blockers in Slack to get a consensus if necessary
- [ ] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [ ] Host another pre-design meeting in `#expensify-open-source` to ask for engineering feedback on the technical solution.
- [ ] Fill out the **Detailed implementation of the solution** and related sections.
- [ ] Re-add the `DesignDocReview` label to this issue
- [ ] Respond to any questions or concerns and bring up blockers in Slack to get consensus if necessary
- [ ] Confirm that the doc has the minimum necessary number of reviews before proceeding
- [ ] Email `strategy@expensify.com` one last time to let them know the Design Doc is moving into the implementation phase
- [ ] Implement the changes
- [ ] Add regression tests so that QA can test your feature with every deploy ([instructions](https://stackoverflowteams.com/c/expensify/questions/363))
- [ ] Send out a follow up email to `strategy@expensify.com` once everything has been implemented and do a **Project Wrap-Up** retrospective that provides:
    - Summary of what we accomplished with this project
    - What went well?
    - What could we have done better?
    - What did we learn?

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by Julesssss** [comment]
After discussion [here](https://expensify.slack.com/archives/C04878MDF34/p1719915230925269?thread_ts=1719581747.585579&cid=C04878MDF34), I want to make the document changes much clearer before seeking help from SWM.

**Comment by trjExpensify** [comment]
I think this is going in #wave-control, right @JmillsExpensify?

**Comment by zsgreenwald** [comment]
Back from OOO, @Julesssss am I needed anywhere in particular to keep this moving?

**Comment by Julesssss** [comment]
Hey, it would be helpful to resolve [this](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?pli=1&disco=AAAA_gJJBt4) comment threads, but other than that no, I'll try to hand over the doc to SWM this week.

**Comment by Julesssss** [comment]
@zsgreenwald actually if you had a chance to check the high-level document again today that would be helpful. I tried to update any section that mentioned the old UI (Avatar instead of Name, etc), but there's a chance I might have missed something. 

Then we should be able to open the project up to SWM tomorrow 🤞

**Comment by Julesssss** [comment]
The [design document](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?usp=sharing) for this project is now ready for the external teams to make a start on. 

myself and @zsgreenwald  will be available for any concerns or questions you have while we work on this project. I have added document comments in places that I think are outdated or need changes. Please feel free to resolve these as you work through the document. And to create new comments with any questions you have for us! 

**Plan should be for the SWM team:**

- Read the [high-level section](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?pli=1#bookmark=id.8q5mhfqb60uw) for background on the project
- Review, update and improve the [detailed section](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?pli=1#bookmark=id.uzzc30pp4si9) of the document. Paying attention to:
    - Design changes, mainly the switch from Attendee Pills to a Text list of attendees (screenshots below)
    - Codebase changes and refactors since November which make some document details irrelevant/outdated
- Submit the proposed detailed changes for another round of internal feedback
- Start work on the feature, while I handle the back-end changes

_________________

**Previous Design**

![Screenshot 2024-06-28 at 14 32 00](https://github.com/user-attachments/assets/5c76fd9c-1be3-4fb6-8d06-d45fdfd23f03)

**Updated Design**

![Screenshot 2024-07-18 at 14 49 04](https://github.com/user-attachments/assets/7857ed31-e5f2-43e5-9c33-0063113212f9)

**Comment by filip-solecki** [comment]
Hi! I am Filip from SWM and I'd like to work on this issue!

**Comment by zfurtak** [comment]
Hello, I'm Zuza from SWM and I'd like to work on this issue with Filip :)

**Comment by Julesssss** [comment]
Design Document review/update is in progress. I'll continue to monitor comments and help the SWM team with any questions.

**Comment by zfurtak** [comment]
We have reviewed the doc and got all the answers, from our side it's ready 😊

**Comment by Julesssss** [comment]
Excellent. I just noticed [this comment](https://docs.google.com/document/d/1e6w0uVdCh_h_uNJEBWwi2rmPzTCLKTb-HdimFnVEfqk/edit?disco=AAABRv1CiC8&pli=1), which I think will need to be reviewed. Other than that it looks like we're just held on my backend changes which I'll aim to finish up tomorrow.

**Comment by zfurtak** [comment]
Done 💪

**Comment by Julesssss** [comment]
Excellent. All comments resolved, except [one here](https://expensify.slack.com/archives/C06ML6X0W9L/p1721999989053519) where we have brought a backend proposal into the Wave-Control room. 

On Monday morning I will share the doc -- assuming no issue is raised with our proposal

**Comment by Julesssss** [comment]
I have written the email, but holding on sending until [this](https://expensify.slack.com/archives/C06ML6X0W9L/p1722214490371679?thread_ts=1721999989.053519&cid=C06ML6X0W9L) internal comment/discussion is resolved.

**Comment by Julesssss** [comment]
Actually, Yuwen is 80% OOO, so I'm going to move forward. We can restart the thread discussion if anyone comments about it.

**Comment by Julesssss** [comment]
The updated design doc has been sent out for review.

**Comment by Julesssss** [comment]
We're making progress through the comments. I have spent the afternoon rethinking some things. Tomorrow I will request public feedback on the issue of saving recent attendees: 

- Today we have a limit of 10, which doesn't align with NewDot
- There is no timestamp for these attendees, which might require a bigger refactor than first planned
- Our existing recent attendee logic has bugs that we may/may not care about resolving

**Comment by Julesssss** [comment]
Still waiting on two reviewers, but the NVP discussion I took to slack has been resolved and the doc has been updated. I should be able to start creating and organising the issues tomorrow 🤞 

I will likely create issues starting those without dependencies, tagging them under wave-control and linking them to the tracking issue.

**Comment by Julesssss** [comment]
Here is the issue breakdown. They are being created tomorrow, and I'll start working on backend issues that will allow SWM to receive attendee onyx data.

![Screenshot 2024-08-08 at 16 36 15](https://github.com/user-attachments/assets/c72a7305-5296-4bf1-bb90-88d8a1ebda8a)

**Comment by Julesssss** [comment]
Today I created the issues which you can find at the top of this tracking issue. I broke down each issue into better-organised tasks, and we can further split up the tasks into separate issues if necessary. We have just six to complete: three frontend and three backend. @mjasikowski and I will make a start on the backend issues next week.

![Screenshot 2024-08-09 at 16 30 36](https://github.com/user-attachments/assets/8f2160a4-f5d0-4b33-a1b5-f24b57fd0389)

**Comment by JmillsExpensify** [comment]
Latest updates [in Slack](https://expensify.slack.com/archives/C06ML6X0W9L/p1724093948469269).

**Comment by Julesssss** [comment]
The initial App issue, and the 2nd backend Issues are currently in progress.

**Comment by Julesssss** [comment]
Hey all, today I noticed a flaw in the design doc. Some further changes will be necessary to correctly persisted attendee data to the transaction, and to update the param that we currently use to pass this data from the front-end. Further discussion [here](https://expensify.slack.com/archives/C06ML6X0W9L/p1725026268311649) in Slack.

**Comment by Julesssss** [comment]
Today, I caught up with @dylanexpensify. I now understand the additional changes that will be required, but the deadline is going to be too tight to meet. Rushing to complete the project would require skipping our rollout, testing, and bug-fixing best practices so I have asked for the project to be dropped from this weeks deadline. The main causes are the details I missed in the detailed design doc and splitting my time between HybridApp and AttendeeTracking. However, we have a good plan for finishing up the project within the next couple of weeks.  

**Here is the plan:** 
- @Julesssss starts work on the updated logic for persisting attendee data to the transaction 
- @lakchote finishes up the original backend tasks 
- @lakchote joins me as a project manager and we test/merge PRs in isolation
- @lakchote and I complete the remaining backend issues

**Here are the additional tasks:** 
- Update the Design Doc and get our changes reviewed
- https://github.com/Expensify/Expensify/issues/425201 [PR in review]
- Create 'UpdateMoneyRequestAttendees' command
- Hook up the recent attendee NVP logic to the new changes
- Clear out uses of the now-unused Auth attendee param

**Comment by Julesssss** [comment]
- The initial two front-end code has been merged to our `feature-attendeeTracking` feature branch
- The revised changes have been added to the design doc (highlighted in red)
- The backend additions for attendee transaction persistence is very close to being merged

The next step is to finish the EditMoneyRequest command changes, share onyx updates for attendees and recentAttendees, then get further front-end changes implemented.

**Comment by Julesssss** [comment]
New update in [Slack](https://expensify.slack.com/archives/C06ML6X0W9L/p1725618620476609?thread_ts=1725273762.458159&cid=C06ML6X0W9L).

**Comment by Julesssss** [comment]
Progress is being made on the front-end and backend. We're currently held on merging @lakchote's EditMoneyRequest changes until the next Web-E and Auth deployments.

**Comment by zfurtak** [comment]
I spotted another issue. When selecting-unselecting the submitter `Recents` section disappears and appears. I don't know if we should address it right now? Does the backend return correct recents already? 🤔
Video below:

https://github.com/user-attachments/assets/31e36c44-435e-42a6-873c-b33cc1732438

cc: @Julesssss

**Comment by Julesssss** [comment]
@zfurtak did you notice that against a specific issue, or the feature branch? We're not yet returning attendees as Onyx data, though shouldn't we be seeing recent contacts here as a fallback? (the same that appear under the header when creating a new chat)

**Comment by zfurtak** [comment]
Here I use default `Recents` to test, so I guess we need to check it when the `Recents` data is correct, the never mind 😊

**Comment by Julesssss** [comment]
@lakchote and I caught up today. The front-end Edit attendees PR is in review and thenext step will be to push Onyx data back to the client under `transaction.attendees`. Lucien and I are both working on other tasks, so I will seek an internal engineer to help us with this task early next week. 

While we could optimistically set and use this data on the front-end now, it's going to be much easier to await these backend changes.

**Comment by dylanexpensify** [comment]
@Julesssss did we want to create a GH issue for that and get it as a hotpick?

**Comment by Julesssss** [comment]
@dylanexpensify just holding on [this](https://expensify.slack.com/archives/C07CACYRTJS/p1726481324257989?thread_ts=1726469920.057659&cid=C07CACYRTJS) briefly -- we need to switch focus to a SmartScan issue that arose from our backend changes to EditRequest

**Comment by Julesssss** [comment]
This morning I worked through all of the remaining issues/tasks and have organised the issues. We have three remaining issues, two of which are held on an issue I will need to implement. I hope to start on this task this week (after I am able to resolve a HybridApp fire):

**Frontend:**
- https://github.com/Expensify/App/issues/47171
- https://github.com/Expensify/App/issues/47172
- https://github.com/Expensify/App/issues/47173
- https://github.com/Expensify/App/issues/48695
- https://github.com/Expensify/App/issues/48692

**Backend**
- https://github.com/Expensify/Expensify/issues/425201
- https://github.com/Expensify/Expensify/issues/419205
- https://github.com/Expensify/Expensify/issues/419206
- https://github.com/Expensify/Expensify/issues/419218
- https://github.com/Expensify/App/issues/49326
- https://github.com/Expensify/Expensify/issues/429056

**Comment by Julesssss** [comment]
Small update, but I am working on the blocking backend issue as my top priority.

**Comment by Julesssss** [comment]
Attendee Tracking is currently my top priority, I am making progress with the [backend PR ](https://github.com/Expensify/App/issues/49326)that unlocks the other two remaining issues.

**Comment by Julesssss** [comment]
**Update:** 

I have been making slow but steady progress with the [new Auth command](https://github.com/Expensify/App/issues/49326#issuecomment-2411159383), while testing the full flow via Web to client. It's likely I will be submitting the Auth command PR soon which passes back attendee data and creates the new action -- but will not update the recent attendees NVP. I'll do that later. 

When the Auth PR is merged, Web-E will be next, then I can open up the App issue so that we can adjust the API param names and onyx keys.

**Comment by Julesssss** [comment]
Internal prioritisation means that this project needs to wind down.

Thankfully the past few weeks has put the project in a good place; the new command is mostly implemented and the onyx attendee data is being passed back to the App clients. I will get the PRs merged, and leave detailed notes on what remains to be done here. 

**Finish the project**
- Within `createUpdateMoneyRequestAttendees`, update the users recent attendees and pass back to client
- App client should update the param used to pass attendees to backend
- App client should switch to the `transaction.attendees` onyx keys

**Post-launch polish:**
- Clear out uses of the now-unused Auth attendee param
- Add more tests like testRequestMoneyPersistsAttendees for edit case
- https://github.com/Expensify/App/issues/50802
- Once fully migrated, introduce minimum attendee size of 1:
`Request::verifyAttributeSize(request, "attendees", 0, Request::MAX_SIZE_NONCOLUMN);`

**Comment by zanyrenney** [comment]
When this project is picked back up, mind letting me know? I have a lot of other priorities so I have the help site  document ready but I don't want to sink more time into this if it's on pause for now @Julesssss 🙏🏼 Thanks!

**Comment by Julesssss** [comment]
Yeah, so that I don't forget I assigned myself to your [issue](https://github.com/Expensify/Expensify/issues/427725) and will check in monthly.

**Comment by Julesssss** [comment]
The task list has been updated. The App feature branch is ready to be merged today and is currently in [review](https://github.com/Expensify/App/pull/51329).

**Comment by Julesssss** [comment]
Of course 😄 

Please correct me if I forgot, but I think these are the PRs you reviewed?
- https://github.com/Expensify/App/pull/48008
- https://github.com/Expensify/App/pull/48870
- https://github.com/Expensify/App/pull/49008

**Comment by JmillsExpensify** [comment]
Still on hold

**Comment by garrettmknight** [comment]
Noting that I'm consolidating the internal tracking issue with this one. I've added steps/documentation from the internal OP to this issue.

**Comment by Julesssss** [comment]
Project remains on hold. As a reminder the 'Finish the project' tasks are listed in the issue description and remain accurate.

**Comment by Julesssss** [comment]
In #convert we are[ planning](https://expensify.slack.com/archives/C03TQ48KC/p1736804641790249) to finish up this project. I have shared the details of what tasks remain in engineering-chat and will happily guide any engineer through the remaining tasks!

**Comment by Julesssss** [comment]
That's great, thanks Francois. I'll be happy to help you with any questions you have. `Finish the project` issues still look correct to me.

**Comment by JmillsExpensify** [comment]
Recently picked back up.

**Comment by francoisl** [comment]
No longer on hold, working on this now.

**Comment by JmillsExpensify** [comment]
In implementation

**Comment by francoisl** [comment]
In progress! Aiming for mid-next week to have everything in review.

**Comment by danielrvidal** [comment]
Wow, awesome! This is an exciting one!

**Comment by Julesssss** [comment]
I jumped in to help out with one of the remaining tasks while Francois is away. [Here](https://github.com/Expensify/App/pull/58120#issuecomment-2755738677) we are discussing how to get the attendee data to display correctly within the edited expense reportAction.

_____________________________________

I also noticed a small task that may already be known. It shouldn't be too difficult as we should already be storing recentAttendees in onyx. When editing an expense we highlight recent attendees to the user:

<img width="816" alt="Image" src="https://github.com/user-attachments/assets/d7d7ce2a-02b0-47d9-bdca-fc353dee377c" />

But this section isn't shown when for a new expense: 

<img width="820" alt="Image" src="https://github.com/user-attachments/assets/1fd0858b-aacc-4299-86ef-443ea01070dc" />

**Comment by allroundexperts** [comment]
> I also noticed a small task that may already be known. It shouldn't be too difficult as we should already be storing recentAttendees in onyx. When editing an expense we highlight recent attendees to the user:

@Julesssss This is because `nvp_expensify_recentAttendees` isn't being sent by the backend. I checked the onyx data and this key did not exist.

**Comment by JmillsExpensify** [comment]
What's the latest on this project?

**Comment by allroundexperts** [comment]
Working on polish issues. Draft PR [here](https://github.com/Expensify/App/pull/60067). The PR needs some backend changes in order to become ready for review. @francoisl is handling those.

https://expensify.slack.com/archives/C08CZDJFJ77/p1744292936926739

https://expensify.slack.com/archives/C08CZDJFJ77/p1744308271762649

**Comment by allroundexperts** [comment]
Polish issues status:

1. https://github.com/Expensify/App/issues/59600 -> 🟡 (https://github.com/Expensify/App/pull/60067)
2. https://github.com/Expensify/App/issues/59599 -> 🟡 (https://github.com/Expensify/App/pull/60067) (Discussion https://expensify.slack.com/archives/C08CZDJFJ77/p1744292936926739)
3. https://github.com/Expensify/App/issues/59597 -> 🟡 (https://github.com/Expensify/App/pull/60067)
https://github.com/Expensify/App/issues/59631 -> 🟡 (https://github.com/Expensify/App/pull/60067) Requires backend fixes as well (https://expensify.slack.com/archives/C08CZDJFJ77/p1744292936926739)
https://github.com/Expensify/App/issues/59629 -> 🔵 
https://github.com/Expensify/App/issues/59621 -> 🟡 (https://github.com/Expensify/App/pull/60067)
https://github.com/Expensify/App/issues/59617 -> 🟡 (https://github.com/Expensify/App/pull/60067)
https://github.com/Expensify/App/issues/59606 -> Probably not a bug ([Discussion](https://expensify.slack.com/archives/C08CZDJFJ77/p1744308271762649))
https://github.com/Expensify/App/issues/59603 -> 🟡 (https://github.com/Expensify/App/pull/60067)
https://github.com/Expensify/App/issues/59602 -> 🟡 (https://github.com/Expensify/App/pull/60067)

🟡 -> In progress
🔵 -> Not touched yet

**Comment by allroundexperts** [comment]
Update: Still waiting on the backend changes

**Comment by allroundexperts** [comment]
Update: Backend changes are deployed. Working on validating those.

**Comment by JmillsExpensify** [comment]
No assigned users require payment, but I also don't think we're ready to close yet as this initiative is not yet done.

**Comment by mananjadhav** [comment]
Looks like the PR I reviewed hasn't been deployed to production yet?

**Comment by francoisl** [comment]
@JmillsExpensify I have a WIP issue [here](https://github.com/Expensify/Expensify/issues/504765), but holding until we finish making Attendee Tracking a feature, so we can also add tests for that.

**Comment by JmillsExpensify** [comment]
@mananjadhav is going to confirm that #61484 is deployed to production, though ahead of that confirmation, here's the payment summary: $500 to @mananjadhav for [PR review and testing](https://github.com/Expensify/App/issues/44725#issuecomment-2909961767).

**Comment by francoisl** [comment]
Yes, that looks correct to me.

**Comment by JmillsExpensify** [comment]
Payment summary: $500 to @mananjadhav for PR review and testing.

**Comment by garrettmknight** [comment]
$500 approved for @mananjadhav

**Comment by JmillsExpensify** [comment]
I think we're done here though please re-open if something is missed.
